trigger/project_triggers.go

```diff
@@ -0,0 +1,116 @@
+package trigger
+
+import (
+	"context"
+	"encoding/base64"
+	"time"
+
+	"github.com/evergreen-ci/evergreen"
+	"github.com/evergreen-ci/evergreen/model"
+	"github.com/evergreen-ci/evergreen/model/user"
+	"github.com/evergreen-ci/evergreen/model/version"
+	"github.com/evergreen-ci/evergreen/repotracker"
+	"github.com/evergreen-ci/evergreen/thirdparty"
+	"github.com/pkg/errors"
+)
+
+// TriggerDownstreamVersion assumes that you definitely want to create a downstream version
+// and will go through the process of version creation given a triggering version
+func TriggerDownstreamVersion(args ProcessorArgs) (*version.Version, error) {
+	if args.Command != "" {
+		return nil, errors.New("command-based triggers are not yet implemented")
+	}
+	if args.File != "" && args.Command != "" {
+		return nil, errors.New("cannot specify both a file and command")
+	}
+	if args.SourceVersion == nil {
+		return nil, errors.Errorf("unable to find source version in project %s", args.DownstreamProject.Identifier)
+	}
+
+	// propagate version metadata to the downstream version
+	metadata, err := metadataFromVersion(*args.SourceVersion, args.DownstreamProject)
+	if err != nil {
+		return nil, err
+	}
+	metadata.TriggerID = args.TriggerID
+	metadata.TriggerType = args.TriggerType
+	metadata.EventID = args.EventID
+	metadata.DefinitionID = args.DefinitionID
+
+	// get the downstream config
+	config, err := makeDownstreamConfigFromFile(args.DownstreamProject, args.File)
+	if err != nil {
+		return nil, err
+	}
+
+	// create version
+	v, err := repotracker.CreateVersionFromConfig(&args.DownstreamProject, config, metadata, false, nil)
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating version")
+	}
+	err = model.UpdateLastRevision(v.Identifier, v.Revision)
+	if err != nil {
+		return nil, errors.Wrap(err, "error updating last revision")
+	}
+	err = repotracker.AddBuildBreakSubscriptions(v, &args.DownstreamProject)
+	if err != nil {
+		return nil, errors.Wrap(err, "error adding build break subscriptions")
+	}
+
+	return v, model.DoProjectActivation(args.DownstreamProject.Identifier)
```

Comment:
do we want to wrap the error? does it make sense to return a non-nil version, with a non-nil error

Reply:
Changed. In general though I think the expectation is that returning a non-nil error means the other return values should be disregarded